04-vue-cli/02-TheToaster/components/TheToast.vue

```diff
@@ -0,0 +1,100 @@
+<template>
+  <div :class="['toast', data.className]">
+    <UiIcon class="toast__icon" :icon="data.icon" />
+    <span>{{ text }}</span>
+  </div>
+</template>
+
+<script lang="ts">
+import UiIcon from './UiIcon.vue';
+import { defineComponent } from 'vue';
+
+export enum EToastKind {
+  SUCCESS = 'SUCCESS',
+  ERROR = 'ERROR',
+}
+
+type TToastData = {
+  icon: string;
+  className: string;
+};
+
+export default defineComponent({
+  name: 'TheToast',
+
+  components: { UiIcon },
+
+  props: {
+    kind: {
+      type: String,
+      default: EToastKind.SUCCESS,
+      validator(value: string): boolean {
+        return value in EToastKind;
+      },
+    },
+
+    text: { type: String, required: true },
+
+    timeout: { type: Number, default: 5000 },
+  },
+
+  computed: {
+    data(): TToastData {
+      switch (this.kind) {
+        case EToastKind.ERROR:
+          return {
+            icon: 'alert-circle',
+            className: 'toast_error',
+          };
+
+        default:
+        case EToastKind.SUCCESS:
+          return {
+            icon: 'check-circle',
+            className: 'toast_success',
+          };
+      }
+    },
+  },
+
+  emits: ['close'],
+
+  mounted() {
+    setTimeout(() => {
+      this.$emit('close', this.text);
+    }, this.timeout);
+  },
```

Comment:
Passing `this.text` with the event is redundant.
1. This text was literally got from the parent. `TheToast` takes `text` from the parent with a prop and then sends it back to the parent. The parent knows the text itself.
2. No way the toast will send another value. "Toast A" cannot emit an event about closing "Toast B". It is always about self-closing.

Reply:
2. Should `TheToast` be a sefl-closing component? In this way `TheToaster`'s data won't be in sync with its actual childs.

Reply:
By "self-closing" I meant that a specific component instance always triggers closing only about itself. That "Toast A" can only ask the parent to close "Toast A", not "Toast B".

But closing should be implemented by removing the data in the parent anyway.